### PR TITLE
Backend tests for file upload endpoints

### LINKED ISSUE #1165: [Backend tests] File upload
Write tests that cover scenarios for each type of campaigns. 
- a user uploading recipient file
- server retrieves the uploaded file from S3
- processing the file

The endpoints involved are `/upload/start`, `/upload/complete` and `/upload/status`

**Possible test scenarios**
- success/error getting pre-signed url
- valid and invalid file uploads during file processing
- upload timeout exceeded when checking file upload status
- file upload failing to other errors 

**Note:** We can split this task up to work on separate PRs for each channel

### LINKED ISSUE #1166: [Backend tests] Substituting templates and message sanitisation
Write tests that cover scenarios for
- substitution of template after recipient file is processed
- sanitisation of messages before insertion to message table

**Possible test scenarios**
- template variables are substituted correctly
- script/html tags are filtered out and message is sanitised 
- only allowed set of html tags present in sanitised message (different for each channel type)

### PR DESCRIPTION
**Pull request by miazima**
## Problem

Closes #1165 and closes #1166

## Solution

- Add tests for file upload and upload status routes for email, protected email and telegram channels

### CODE REVIEW

**Review by lamkeewei — CHANGES_REQUESTED** [review]
@miazima - we need to also test the different scenarios of the file parsing successes/errors as stated in #1165.

Some of the missing scenarios are:
- Invalid recipient list with missing params
- Ensure that messages are populated correctly in the respective messages table from uploading a valid recipient list

For these cases, we will need to mock the file download from S3.

**Comment by miazima** [comment]
I have added on missing scenarios, however the tests involving the file processing in `...-template.middleware.test.ts` are failing intermittently, am trying to figure what is causing it

File: backend/src/sms/middlewares/tests/sms-template.middleware.test.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      mockRequest as Request,
+      mockResponse as Response,
+      nextFunction
+    )
+
+    await new Promise<void>((resolve) =>
+      setTimeout(() => {
+        resolve()
+      }, 1000)
+    )
```
Comment (miazima) [inline backend/src/sms/middlewares/tests/sms-template.middleware.test.ts:132]: The following asserts on the sms messages table being populated fails, I suspect due to the async `parseAndProcessCsv` not having completed running. However this sleep function does not seem to resolve the issue. Similar issue for the telegram template middleware.

File: backend/src/core/services/redis.service.ts [hunk 2]
```diff
@@ -50,7 +50,31 @@ const sessionClient = redis
...
+      resolve()
+    })
+  })
+
+  await Promise.all([closeOtpClient, closeSessionClient])
+
+  // redis.quit() creates a thread to close the connection.
+  // We wait until all threads have been run once to ensure the connection closes.
+  await new Promise((resolve) => setImmediate(resolve))
+}
```
Comment (miazima) [inline backend/src/core/services/redis.service.ts:74]: Added this close function to wait for the redis connections to close, this did not resolve the open handles issue, unless the implementation is flawed?

**Comment by lamkeewei** [comment]
To close in favour of #1202

**Comment by stanleynguyen** [comment]
Outdated due to #1202 implementation

**Pull request closed without merging**